typed-function.js

```diff
@@ -1259,96 +1259,137 @@
+    }
 
-      for (var i = 0; i < fns.length; i++) {
-        var fn = fns[i];
-
-        // test whether this is a typed-function
-        if (typeof fn.signatures === 'object') {
-          // merge the signatures
-          for (var signature in fn.signatures) {
-            if (fn.signatures.hasOwnProperty(signature)) {
-              validateUnique(signature, fn.signatures[signature]);
-              signaturesMap[signature] = fn.signatures[signature];
+    /**
+     * Copy all of the signatures from the second argument into the first,
+     * which is modified by side effect, checking for conflicts
+     */
+    function mergeSignatures (dest, source) {
+      for (const key in source) {
+        if (source.hasOwnProperty(key)) {
+          if (key in dest) {
+            if (source[key] !== dest[key]) {
+              const err = new Error('Signature "' + key + '" is defined twice')
+              err.data = { signature: key }
+              throw err
             }
+            // else: both signatures point to the same function, that's fine
           }
-        }
-        else if (typeof fn.signature === 'string') {
-          validateUnique(fn.signature, fn);
-          signaturesMap[fn.signature] = fn;
-        }
-        else {
-          err = new TypeError('Function is no typed-function (index: ' + i + ')');
-          err.data = {index: i};
-          throw err;
+          dest[key] = source[key]
         }
       }
-
-      return signaturesMap;
     }
 
-    typed = createTypedFunction('typed', {
-      'string, Object': createTypedFunction,
-      'Object': function (signaturesMap) {
-        // find existing name
-        var fns = [];
-        for (var signature in signaturesMap) {
-          if (signaturesMap.hasOwnProperty(signature)) {
-            fns.push(signaturesMap[signature]);
+    /**
+     * Originally the main function was a typed function itself, but then
+     * it might not be able to generate error messages if the client
+     * replaced the type system with different names.
+     *
+     * Main entry: typed([name], functions/objects with signatures...)
+     *
+     * Assembles and returns a new typed-function from the given items
+     * that provide signatures and implementations, each of which may be
+     * * a plain object mapping (string) signatures to implementing functions,
+     * * a previously constructed typed function, or
+     * * any other single function with a string-valued property `signature`.
+
+     * The name of the resulting typed-function will be given by the
+     * string-valued name argument if present, or if not, by the name
+     * of any of the arguments that have one, as long as any that do are
+     * consistent with each other. If no name is specified, the name will be
+     * an empty string.
+     *
+     * @param {string} name [optional]
+     * @param {(function|object)[]} signature providers
+     * @returns {typed-function}
+     */
+    typed = function(maybeName) {
```

Comment:
I will try this and see if it works the same way; my main concern was whether you can "replace" an existing variable with a classic-style function definition. Will let you know how it goes.

Reply:
Yeah, if one makes this change, at the very first call to typed() in a test, JavaScript throws a `TypeError: typed is not a function` error. So I think that since typed is first a temporary non-function object, there's no choice but to reassign it with an anonymous function, and it seems to work. If it fails a particular browser let me know and maybe we can define a named function with a different name and assign that to typed and then change the name property of that new function to 'typed' but that seems more convoluted if an anonymous function works.

Reply:
>  So I think that since typed is first a temporary non-function object, there's no choice but to reassign it

Ah, yes I understand. Ok let's leave it like this. It would be nice to see if we can get rid of this temporary non-function object somehow, but better to try that in a future PR.